src/dns_resolver_ares.c

```diff
@@ -0,0 +1,283 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+// This file is OS-specific, and is identified by setting include directories
+// in the project
+#include "socket_async_os.h"
+
+#include "azure_c_shared_utility/dns_resolver.h"
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/crt_abstractions.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "ares.h"
+
+// EXTRACT_IPV4 pulls the uint32_t IPv4 address out of an addrinfo struct
+// This will not be needed for the asynchronous design
+#ifdef WIN32_NOT_USED
+//#define EXTRACT_IPV4(ptr) ((struct sockaddr_in *) ptr->ai_addr)->sin_addr.S_un.S_addr
+#else
+// The default definition handles lwIP. Please add comments for other systems tested.
+#define EXTRACT_IPV4(ptr) ((struct sockaddr_in *) ptr->ai_addr)->sin_addr.s_addr
+#endif
+
+typedef struct
+{
+    char* hostname;
+    int port;
+    uint32_t ip_v4;
+    bool is_complete;
+    bool is_failed;
+    bool in_progress;
+    struct addrinfo* addrInfo;
+    ares_channel ares_resolver;
+} DNSRESOLVER_INSTANCE;
+
+DNSRESOLVER_HANDLE dns_resolver_create(const char* hostname, int port, DNSRESOLVER_OPTIONS* options)
+{
+    /* Codes_SRS_dns_resolver_30_012: [ The optional options parameter shall be ignored. ]*/
+    DNSRESOLVER_INSTANCE* result;
+    int status;
+    (void)options;
+    if (hostname == NULL)
+    {
+        /* Codes_SRS_dns_resolver_30_011: [ If the hostname parameter is NULL, dns_resolver_create shall log an error and return NULL. ]*/
+        LogError("NULL hostname");
+        result = NULL;
+    }
+    else
+    {
+        result = malloc(sizeof(DNSRESOLVER_INSTANCE));
+        if (result == NULL)
+        {
+            /* Codes_SRS_dns_resolver_30_014: [ On any failure, dns_resolver_create shall log an error and return NULL. ]*/
+            LogError("malloc instance failed");
+            result = NULL;
+        }
+        else
+        {
+            int ms_result;
+            result->is_complete = false;
+            result->is_failed = false;
+            result->in_progress = false;
+            result->ip_v4 = 0;
+            result->port = port;
+            /* Codes_SRS_dns_resolver_30_010: [ dns_resolver_create shall make a copy of the hostname parameter to allow immediate deletion by the caller. ]*/
+            ms_result = mallocAndStrcpy_s(&result->hostname, hostname);
+            if (ms_result != 0)
+            {
+                /* Codes_SRS_dns_resolver_30_014: [ On any failure, dns_resolver_create shall log an error and return NULL. ]*/
+                LogError("dns_resolver_create: hostname allocation failed");
+                free(result);
+                result = NULL;
+            }
+
+            status = ares_library_init(ARES_LIB_INIT_WIN32);
+            if (status != ARES_SUCCESS)
```

Comment:
Result will fail and then throw an exception on line 96.  Need to put an else after the if block.